### DISCUSSION TITLE
MIDI2Hydrogen  - A new tool for conversion of MIDI files into Hydrogen songs

### DISCUSSION THREAD

**Discussion by DanielCech** [hydrogen-music/hydrogen#1960] [General]
Hey guys, 

I'd like to offer you to use my hobby project - **MIDI2Hydrogen** converter - https://github.com/DanielCech/Midi2Hydrogen. It is a free macOS utility for conversion of MIDI files into h2song format. I like Hydrogen but I feel there are not enough resources available online - as far as I know. Maybe I am wrong. I don't fully understand the reason why the import of MIDI files is not part of Hydrogen itself. I created this utility primarily for my use for conversion: Guitar Pro > MIDI > Hydrogen. I hope it can be useful also for someone else.

MIDI2Hydrogen important feature is that it detects duplicated patterns and creates the pattern sequence.

**Comment by DanielCech** [comment]
![image](https://github.com/hydrogen-music/hydrogen/assets/3081522/0bee22c7-7723-4156-b06f-404ee4873b05)

**Comment by theGreatWhiteShark** [comment]
Hey @DanielCech ,

Thanks for you work! I added it to our [list of 3rd party software](https://github.com/hydrogen-music/hydrogen/wiki/external-software-&-utilities) in our wiki.

> I like Hydrogen but I feel there are not enough resources available online

In terms of documentation and tutorials or drumkits, songs, and patterns that get you started?

> I don't fully understand the reason why the import of MIDI files is not part of Hydrogen itself.

Well, judging from the code base MIDI support was driven by 1-2 individuals with a very clear use case in mind. But some parts have fallen into despair. E.g. there is MIDI export. But it seems to be written prior to the introduction of the timeline and its tempo markers and was not updated back then. So, no tempo changes are reflected. Neither is humanization of notes. And that for at least 16 years by now!

That's sad. But things are getting better. When I took over as maintainer I told people with MIDI issues that there is no one on the team capable of fixing them. But since then I introduced myself into a couple of things and already fixed a number of bugs in MIDI in and output handling. MIDI import and export is definitely on the TODO list, along with loads of other stuff.

If you would consider contributing MIDI import, this patch would be more than welcome! And I would be happy to support you.

**Comment by DanielCech** [comment]
Hey @theGreatWhiteShark, thank you. Maybe you can add also https://github.com/brandjamie/midi2hydro_pattern that inspired my solution. I am glad I could help.
Regarding the available resources - I can imagine there are a lot of h2songs and h2patterns on the internet available for download but I've found almost nothing. I am not a drummer, I play guitar and bass and my initial idea was to have a rhythm pedal like BeatBuddy - https://www.thomann.de/gb/singular_sound_beatbuddy.htm. But it is way too expensive. So I created a solution based on my custom MIDI controller and Hydrogen on Raspberry Pi. And I am very satisfied with the result.

![IMG_8719](https://github.com/hydrogen-music/hydrogen/assets/3081522/ee5509c0-4dda-4d9c-a528-02e7580f1d1c)
![C2541B13-8E2E-49F6-B9E3-1F9DDE3B8ECE_1_105_c](https://github.com/hydrogen-music/hydrogen/assets/3081522/e21e1b0c-0aaa-46b0-b6a5-4cb88a792287)

It would be great to have a collection of royalty-free rhythms and patterns for Hydrogen. But I've found only fragments from various resources.

Of course, I'd love to see this feature in the Hydrogen app. Unfortunately, I am very busy so I am not sure what I can offer you as a collaboration. I appreciate your offer. Of course, my utility is very basic and you're looking for a more profound solution. Let's stay in contact.

**Reply by theGreatWhiteShark** [reply]
Nice setup!

> It would be great to have a collection of royalty-free rhythms and patterns for Hydrogen. But I've found only fragments from various resources.

In Hydrogen patterns are written for a particular drumkit. When switching to another kit the notes of the first instrument of the former kit are mapped to the first instrument of the new one, the second instrument to the second, and so forth. This means you have to carefully fine tune your kits in order to switch between them without the need to rearrange notes. Or you have some local scripts taking care of this.

For version 1.3. I intend to fix this design flaw and allow for Hydrogen to seamlessly switch between kits. (That's why it will probably still take quite a while until the release is ready and well tested).

This missing portability is probably the main reason there are so little songs and patterns being shared via internet. Let's see, maybe this will change during the next years.

**Reply by DanielCech** [reply]
@theGreatWhiteShark  I see. Makes sense. Thanks for the explanation. I am looking forward to the new version with this functionality.
I noticed a bug in my utility - the instrument mapping is not working great. I'll try to come up with the solution in the next version soon. Probably a JSON file with the mapping of MIDI events to the particular drum set's instrument. I'll use this notation cheat sheet from GuitarPro:

![image](https://github.com/hydrogen-music/hydrogen/assets/3081522/d9b7387d-30f4-4202-9b14-40d1150ca080)

**Reply by theGreatWhiteShark** [reply]
> Probably a JSON file with the mapping of MIDI events to the particular drum set's instrument.

But keep in mind that there are different mapping mechanisms in Hydrogen (see http://hydrogen-music.org/documentation/manual/manual_en.html#chpt.midi.note). There are drumkit-dependent and may differ with settings done in the preferences.

A number of kits try to stick with the [General MIDI standard](https://en.wikipedia.org/wiki/General_MIDI) and have more or less the same mappings. But this is only applicable to smaller ones.

**Reply by DanielCech** [reply]
@theGreatWhiteShark Just a question - where can I switch between `order-based` and `output-based` mapping? I couldn't find it in Preferences or the song file itself. Thanks for the explanation.

**Reply by theGreatWhiteShark** [reply]
> where can I switch between order-based and output-based mapping?

The option is in Preferences > MIDI tab and is called "Use output note as input note". (I know, not the most intuitive name. But it was already there when we entered the project and we are a little conservative when it comes to breaking existing UX).

**Comment by trebmuh** [comment]
There is this: https://github.com/hydrogen-music/Song-and-pattern-repository
